pkg/git/git.go

```diff
@@ -0,0 +1,78 @@
+// Copyright The Shipwright Contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package git
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-git/go-git/v5/config"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+	"github.com/go-git/go-git/v5/storage/memory"
+	"github.com/shipwright-io/build/pkg/ctxlog"
+	"regexp"
+
+	gogitv5 "github.com/go-git/go-git/v5"
+)
+
+const (
+	defaultRemote = "origin"
+	privateRepo   = "private"
+	publicRepo    = "public"
+	invalidRepo   = "invalid"
+)
+
+var (
+	// URLRegexFormat matches the url of git repository
+	httpURLRegexFormat        = regexp.MustCompile(`^((http:\/\/)|(https:\/\/))?([a-zA-Z0-9]([a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,6}(\/)`)
+	sshURLRegexFormat         = regexp.MustCompile(`(ssh:\/\/[\w\d\.]+|.+@?.+\..+:)(:[\d]+){0,1}\/*(.*)`)
+	httpsPublicURLRegexFormat = regexp.MustCompile(`(https:\/\/)?(github\.)+com(\/)`)
+	sshPublicURLRegexFormat   = regexp.MustCompile(`(git+@github.com:)(:[\d]+){0,1}/*(.*)`)
```

Comment:
Are these two URLs really github specific? If yes, then please add some code comment that explains this.

Reply:
@SaschaSchwarze0 I think so, these two URLs are really public github specific. I will add comments later.

Reply:
Done!

Reply:
No, they are not Github specific. Thanks for updating the unit tests. Made it easy to play around with it. Simple gitlab URLs were failing already. Here is a pull request into your branch that changes the regular expressions and includes a test for GitLab: [Identify also non-github URLs #1](https://github.com/xiujuan95/build/pull/1).

Reply:
@SaschaSchwarze0 Merged, thanks a lot!

Reply:
BTW, as you put URLRegex:
```
	publicURLRegex = regexp.MustCompile("https?://[\\w\\.\\:/\\-~]+(\\.git)?/?")

	privateURLRegex = regexp.MustCompile("((git|ssh)|(git@[\\w\\.]+))(:(//)?)[\\w\\.@\\:/\\-~]+(\\.git)?/?")
```
an URL like `https://github.ibm.com` will be treated as a public repo. I think it's not correct, right?

But to be honest, until now, I find I don't know how to distinguish an URL is private or public only by the URL format.
Do you have other ideas about this? Or if we don't need to distinguish private or public, just check it whatever it's private or public?

Reply:
I don't think you can, https://github.com/bigkevmcd/docs.git would come out as public with your matching, but it's a private repository.

From reading the tests, I think you're conflating two things, [private repositories](https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/setting-repository-visibility) with requiring authentication to clone the repository.

For example:

```Go
privateSourceURL6 = "ssh://github.com/shipwright-io/build.git"
```

Clearly, that's not a private repository, because you have...

```Go
publicSourceURL2 = "https://github.com/shipwright-io/build.git"
```

So, the repository `shipwright-io/build` is public (I'm commenting on it just now, and I can clone the source).

Ultimately, the only test of a "private" repository, is whether or not you Git can clone the URL without authentication, URLs and Hosts don't really indicate.

Reply:
Okay, maybe we should change the labeling of the variables and the constants. I was not doing this because I did not want to change too much. In general, I think we have a different distinction: we have HTTP(S) based URLs and we have SSH based URLs.

At the moment, we can only verify HTTP(S) URLs. `https://github.ibm.com/org/repo` is a perfectly valid repository URL and it can also be used within a Build. But, we - and only we - know that this is a GHE URL that is only reachable with an IBM intranet login. The code will detect it as `public` (should maybe rather be HTTP), we then run the validation and will fail. This is the desired behavior because the URL cannot be reached unauthenticated.

Reply:
@bigkevmcd Thanks! But I don't want to clone it, it may degrade our build controller performance, because it will cone the source code when it verify source. Actually, it's unnecessary, right?

Do you have other ideas about verify url is private or public?

Reply:
I don't think it's a useful distinction between "private" and "public", I think "requires authentication", and "doesn't require authentication" is more useful, "private" and "public" have specific meanings in the git hosting sense.

In saying that, as I've indicated elsewhere, if you're not going to actually check the endpoint, then you could identify URLs that will _definitely_ require authentication (i.e. ones with the `ssh` protocol, by parsing the endpoint using go-git's endpoint parser, and checking the `Protocol` field.

Reply:
Thanks for the feedback @bigkevmcd @SaschaSchwarze0 . The initial requirement for this PR was to only validate for public github repositories(as seen in #464) and avoid doing a validation for private ones, or more accurate, for repositories where authentication is required.

In order to avoid doing a validation on repositories that require authentication, we should only check the url(_https valid url_) plus the non-existence of a referenced [secret](https://github.com/shipwright-io/build/blob/65bbdb3ae5491b83810f255b838e2ebcc153bcae/pkg/apis/build/v1alpha1/gitsource.go#L35) for the `source.url`. A different combination of this two parameters should be ignored and not validated.

Everything else that is in the form of `git` or `ssh` as seen in  https://regex101.com/r/CdiB6H/2 should be avoided for now. We will indeed need in the future a validation for urls of the type `git` or `ssh` as long as the referenced [secret](https://github.com/shipwright-io/build/blob/65bbdb3ae5491b83810f255b838e2ebcc153bcae/pkg/apis/build/v1alpha1/gitsource.go#L35) exists. But because we will need to deal with some sort of authentication mechanism (.e.g. consuming the secret data, passing that somewhere), this is out of the scope for now.

Reply:
Im not sure if we will ever do a validation for a private repo, because we are not running this code on a container where we can easily mount the related ssh secrets. In the other hand, we know that Tekton will do this sort of validation for us during a BuildRun "runtime".

I think this PR goal is to prevent users(_as much as possible_) from triggering a BuildRun that will eventually fail because of bad references to a GIT repository(_typos when defining a Build_). We might not be able to cover all scenarios(_private repos_), but at least it help users in the future.